### ISSUE TITLE
Windows + UNC paths

### ISSUE THREAD

**Issue by erdnuesse** [DeaDBeeF-Player/deadbeef#2413]
Hi - not sure if issue or feature missing.
But since I expected it to work - rather an issue.

### Steps to reproduce the problem
Drag and drop a file or folder from a UNC network share into the list. 
OR
Select a file via the add file menu and browse to a UNC path.

### What's going on? Describe the problem in as much detail as possible.
Result: nothing gets added
Win has a security mechanism in place, preventing from eg. copying UNC files to the desktop: 
I already took steps to rule out the problem of windows security. (add unc to intranet manually)

I then added a Network Drive (eg. Z:\) mounting said UNC path, and here goes, all functions properly. 
So there must be a different file handling, and not done via OS calls (not a developer, just trying my best to describe here)

Tl;dr: playing via win network paths do not work, except when adding a network drive with drive letter. 

### Information about the software:

Deadbeef version: latest devel:  SHA1: a18c62cab9dd5a75f197cbab781fddaa3234362d
OS: Win 10 latest.

**Comment by Alexey-Yakovenko** [comment]
@kuba160 FYI

**Comment by kuba160** [comment]
I assume that adding UNC paths through file browser `File/Add file(s)` don't work as well? To make deadbeef compatible with windows paths, they all get converted and backward slash `\` is replaced with forward slash  `\`. I don't use network share, so I didn't have the possibility to test them. I looked at code and in theory UNC paths should get converted back to use backward slash when loading file. Does your path to the file include any special characters? Could you send the path that causes problems? If I find anything wrong with logic in deadbeef then I might be able to fix it. Otherwise I can provide you a build that could provide more information why this fails.

**Comment by erdnuesse** [comment]
See above - yes. 
a)
That's what I meant when I mentioned:
> OR
> Select a file via the add file menu and browse to a UNC path.

b)
No special chars, and due to privacy reasons obviously can't provide exact paths without obfuscation - but that's basically missing the point... right?

c) it's the same path when it got basically fixed / worked-around by adding the drive letter, so why investigate in an obviously wrong direction? 

d) Drive letters are good for a quick fix, but no permanent solution for me, that's why this issue was raised in the first place.

**Comment by kuba160** [comment]
> c) it's the same path when it got basically fixed / worked-around by adding the drive letter, so why investigate in an obviously wrong direction?

There were some changes in deadbeef that were needed for paths with drive letters to work. My first guess is that UNC paths do not work as these changes were not made with UNC in mind. These paths are later passed to syscall and as windows docs state:

>fopen accepts paths that are valid on the file system at the point of execution; fopen accepts UNC paths and paths that involve mapped network drives as long as the system that executes the code has access to the share or mapped drive at the time of execution. When you construct paths for fopen, make sure that drives, paths, or network shares will be available in the execution environment. You can use either forward slashes (/) or backslashes (\) as the directory separators in a path.

I made a debug build (https://github.com/DeaDBeeF-for-Windows/deadbeef/actions/runs/174052623) which you can download here: https://github.com/DeaDBeeF-for-Windows/deadbeef/releases/download/2020-07-06/windows-debug-1d648bec9.zip. (Copying deadbeef.exe to your build might also work). This might give more information what is going on. It will look like this:

```
plt_insert_dir_int: dirname: /C:/Users/Public/Music/Sample Music/Kalimba.mp3
plt_insert_file_int: dirname_conv: C:/Users/Public/Music/Sample Music/Kalimba.mp3
plt_insert_file_int: fname: /C:/Users/Public/Music/Sample Music/Kalimba.mp3
plt_insert_file_int: fname_conv: C:/Users/Public/Music/Sample Music/Kalimba.mp3
fopen: filename: C:/Users/Public/Music/Sample Music/Kalimba.mp3
fopen: filename_c: C:\Users\Public\Music\Sample Music\Kalimba.mp3
fopen: filename: C:/Users/$NAME/Desktop/DEADBE~3/config/playlists/0.dbpl.tmp
fopen: filename_c: C:\Users\$NAME\Desktop\DEADBE~3\config\playlists\0.dbpl.tmp
```

**Comment by kuba160** [comment]
I kinda see where the problem can be. In `plt_insert_file_int` the slash gets removed with drive letters but I don't think it will get removed for UNC paths. If fopen filename path starts with 3 slashes then this might cause problems.

**Comment by erdnuesse** [comment]
so in my IT career unc paths always started with: \\server\share 
but since it says explicitly slashes - //server/share is also valid. 
Also as I don't know how you build your paths it should only cut off leading slashes (/ or \) if there is no <letter>+: in the path. it "should" be UNC otherwise. 
But thats just me, thinking out loud. 
Downloading debug now, will report later.

**Comment by erdnuesse** [comment]
Tested. Still same behavior with drag and drop:
Disclaimer: I'm running the app from /users/public/apps so don't be surprised below.
debug:

> plt_insert_dir_int: dirname: //storage/music/#genre/artist/The Album/02 - title.mp3
> plt_insert_file_int: dirname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
> plt_insert_file_int: fname: //storage/music/#genre/artist/The Album/02 - title.mp3
> 
> (deadbeef:12932): GLib-GIO-CRITICAL **: 22:08:05.803: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
> plt_insert_file_int: fname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
> fopen: filename: file:////storage/music/#genre/artist/The Album/02 - title.mp3
> fopen: filename_c: file:\\\\storage\music\#genre\artist\The Album\02 - title.mp3
> fopen: wfopen failed
> fopen: filename: C:/Users/Public(deadbeef:1/2932): GLib-GIO-CRITICAL **: 22:08:05.824: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
> apps/WINDOW\~1/config/playlists/0.dbpl.tmp
> fopen: filename_c: C:\Users\Public\apps\WINDOW~1\config\playlists\0.dbpl.tmp


Same behavior, different errors when using the file open dialogue though:
> plt_insert_file_int: fname: \\\storage\music\#genre\artist\The Album\02 - title.mp3
> plt_insert_file_int: backslash(es) detected: \\\storage\music\#genre\artist\The Album\02 - title.mp3
> plt_insert_file_int: fname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
> 
> f(deadbeef:12932): GLib-GIO-CRITICAL **: 22:14:58.763: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
> open: filename: C:/Users/Public/apps/WINDOW\~1/config/playlists/0.dbpl.tmp
> fopen: filename_c: C:\Users\Public\apps\WINDOW\~1\config\playlists\0.dbpl.tmp
> fopen: filename: C:/Users/Public/apps/WINDOW\~1\config/playlists/0.dbpl.tmp
> fopen: filename_c: C:\Users\Public\apps\WINDOW\~1\config\playlists\0.dbpl.tmp
> 
> (deadbeef:12932): GLib-GIO-CRITICAL **: 22:14:58.790: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed

(i hope all slashes are copied correctly, this messy editor uses them as escape chars even if formatted as code -.-)
[logcat.txt](https://github.com/DeaDBeeF-Player/deadbeef/files/4944585/logcat.txt)

**Comment by kuba160** [comment]
If you want to format as code you should put text between ``` (if multiline) or single ` (if single line). Then it should work fine.

Anyway from the log you sent:
```
plt_insert_file_int: fname: //storage/music/#genre/artist/The Album/02 - title.mp3
plt_insert_file_int: fname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
fopen: filename: file:////storage/music/#genre/artist/The Album/02 - title.mp3
fopen: filename_c: file:\\\\storage\music\#genre\artist\The Album\02 - title.mp3
fopen: wfopen failed
```
fopen should not get paths with `file://` prefix (at least on windows). This is most likely the cause of the problem. I will need to find what causes it. After that gets fixed, the path sent to syscall will be `\\storage\music\#genre\artist\The Album\02 - title.mp3`. I assume that this is correct path that can be resolved by windows and does not need any adjustments.

**Comment by kuba160** [comment]
I made a version that removes `file://` in fopen: https://github.com/DeaDBeeF-for-Windows/deadbeef/releases/download/2020-07-06/windows-debug-65a2ad712.zip (branch `actions-unc-debug`) Please let me know if it fixes UNC paths for you

**Comment by erdnuesse** [comment]
Same behavior - even downloaded twice, because i was afraid I used the previous one.
Log: (as alwasy, first part - drag 'n' drop, second part - file-open)

```
plt_insert_dir_int: dirname: //storage/music/#genre/artist/The Album/02 - title.mp3
plt_insert_file_int: dirname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
plt_insert_file_int: fname: //storage/music/#genre/artist/The Album/02 - title.mp3
plt_insert_file_int: fname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
f
o(deadbeef:5p920): GLib-GIO-CRITICALm **: 00:28:1n4.587: :g_db us_connfectiion_elmit_signal:e assertion 'G_InS_DBUS_CONNaECTION (connmectieon)' fai:led
 //storage/music/#genre/artist/The Album/02 - title.mp3
fopen: filename_c: \\storage\music\#genre\artist\The Album\02 - title.mp3
fopen: filename: C:/Users/Public/apps/WINDOW~1/
c(deadbeef:5920): GLib-GIO-CoRITICAL0m **: 00:28:14.597: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
fig/playlists/0.dbpl.tmp
fopen: filename_c: C:\Users\Public\apps\WINDOW~1\config\playlists\0.dbpl.tmp

#####

(deadbeef:5920): Gtk-WARNING **: 00:28:19.934: Could not load a pixbuf from icon theme.
This may indicate that pixbuf loaders or the mime database could not be found.
plt_insert_file_int: fname: \\storage\music\#genre\artist\The Album\02 - title.mp3
plt_insert_file_int: backslash(es) detected: \\storage\music\#genre\artist\The Album\02 - title.mp3

(deadbeef:5920): GLib-GIO-CRITICAL **: 00:28:23.839: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
plt_insert_file_int: fname_conv: //storage/music/#genre/artist/The Album/02 - title.mp3
fopen: filename: C:/Users/Public/apps/WINDOW~1/config/playlists/0.dbpl.tmp
fopen: filename_c: C:\Users\Public\apps\WINDOW~1\config\playlists\0.dbpl.tmp

(deadbeef:5920): GLib-GIO-CRITICAL **: 00:28:23.858: g_dbus_connection_emit_signal: assertion 'G_IS_DBUS_CONNECTION (connection)' failed
```

**Comment by kuba160** [comment]
It looks like file got opened correctly (using drag'n'drop) since `fopen: wfopen failed` is missing for newer build. This might mean that the problem is more complex. Interestingly for file/open the file is never fopened. I am currently on vacation and I don't have any windows machine, so I cannot debug the problem now.

**Comment by kuba160** [comment]
So the issue was that the path was detected to be relative, which prevented tracks to get added to the playlist. The `file:\\` prefix was included because it tried to load the file as a playlist and the playlist loading function was not adjusted to support windows. I fixed both issues. If you want to try the debug build, you can download it here: https://github.com/DeaDBeeF-for-Windows/deadbeef/releases/download/2020-07-06/windows-debug-0cdffd618.zip

Changes will be merged soon, after they will be merged you will be able to download normal build from deadbeef's sourceforge page.
